### PR TITLE
Refactor Reputation

### LINKED ISSUE #76: Implement DependencyInjection for JSON deserialization
For some other features, we need our deserialized objects to be injected with their dependencies.

Newtonsoft Json supports this. I found [this example](https://www.newtonsoft.com/json/help/html/DeserializeWithDependencyInjection.htm) that uses AutoFac (another IoC container library) to create the resolver.

> :bulb: Once this is implemented, `LangResource` should get the `Random` class injected that way instead of having it passed in a method.

This might mean going with AutoFac instead of ASP.NET DI. I'm for it if it makes things easier.

### PR DESCRIPTION
**Pull request by petrspelos**
## Related Issue

Fixes #76 
Enables #77 

## Changes
- We had the concept of `Providers` and `Services` under the same `Service` term.
    - `Provider` is something that stands between a DataStructure object and its usecase. It contains all logic related to the DataStructure.
    - `Service` exists to complete a request. This in our case is most frequently executing a command. It interacts with `Providers` to do its job.

- There was a tiny mistake in the self-reputation error message. Fixed that and added a bit of fun into it as well. :yum: 

- I refactored the whole `Reputation` system to use a new `UserReputationProvider`.
    - I updated the tests to now test the Provider and Service separately. The Tests are now cleaner.

## Expected Feedback
Please if someone could just verify I didn't overlook something. :tada: :+1:

### CODE REVIEW

File: src/Miunie.ConsoleApp/Resources/Lang/PhrasesEn.json [hunk 1]
```diff
@@ -23,7 +23,8 @@
     {
         "Key": "CANNOT_SELF_REP",
         "Pool": [
-            "Sorry {0} you can't give rep to yourself!"
+            "Sorry {0} you cannot influence your own reputation! Try influencing other people instead. :smirk:",
```
Comment (svr333) [inline src/Miunie.ConsoleApp/Resources/Lang/PhrasesEn.json]: Sorry {0}, you cannot influence your own reputation!

Just adding a comma would make it neater.

File: src/Miunie.ConsoleApp/Resources/Lang/PhrasesEn.json [hunk 2]
```diff
@@ -23,7 +23,8 @@
...
+            "Aww, {0}... I can't do that. If you need your reputation to change, try blackmailing others to do it for you."
```
Comment (svr333) [inline src/Miunie.ConsoleApp/Resources/Lang/PhrasesEn.json]: I love this

File: src/Miunie.Core/Services/ProfileService.cs [hunk 3]
```diff
@@ -1,50 +1,48 @@
...
             _discordMessages = discordMessages;
-            _userService = userService;
+            _reputationProvider = reputationProvider;
         }
 
         public async Task ShowProfile(MiunieUser u, MiunieChannel c) 
             => await _discordMessages.SendMessage(c, "SHOW_PROFILE", u.Reputation.Value);
 
-        public async Task GiveReputation(MiunieUser source, MiunieUser target, MiunieChannel c)
+        public async Task GiveReputation(MiunieUser invoker, MiunieUser target, MiunieChannel c)
```
Comment (svr333) [inline src/Miunie.Core/Services/ProfileService.cs]: I like the name change, is more correct imo.

**Comment by petrspelos** [comment]
Thanks, @svr333! :wink:

**Pull request merged**